Add Dyn DNS provider

aebruno opened the pull request:
This commit adds support for Dyn, specifically Dyn Managed DNS. Makes
use of the Dyn REST API [1] and requires 3 environment variables for
credentails:

    DYN_CUSTOMER_NAME
    DYN_USER_NAME
    DYN_PASSWORD

The Dyn DNS API requires a session to be established first using the
credentials above, After creating a session an api token is used for
subsequent requests.

The unit test performs a "live test" and currently requires a valid Dyn
account and domain.

[1] https://help.dyn.com/rest/

xenolf commented:
Thank you for this!

I'm sorry that it takes me a while to go over this. We are currently discussing how we want to handle external providers going forward in #100 and #111.

aebruno commented:
Thanks for the great library. I'd be happy to move the dyn provider to a another package or however you end up handling the providers.

xenolf commented:
Hello there and sorry for the wait :worried: 
#144 was merged just now and that's the new layout.

aebruno commented:
Hey, I rebased and moved the Dyn DNS provider out of the acme package per #144. Tried to follow the other providers the best I could. Let me know how things look and if you'd like anything changed.

xenolf commented:
Thanks for your work @aebruno, this is looking quite good. Just one small thing: Could you please use a HTTP client with a timeout like `http.Client{Timeout: 10 * time.Second}` instead of the go default one for your HTTP requests?

aebruno commented:
All set. Added timeout to HTTP client. Let me know if you spot anything else.

aebruno commented:
Not sure why the CI tests failed on looking up the authoritative zone?

xenolf commented on providers/dns/dyn/dyn.go:
```diff
@@ -0,0 +1,249 @@
+
+var dynBaseURL = "https://api.dynect.net/REST"
+
+type DynResponse struct {
```
This doesn't need to be exported, right?

xenolf commented on providers/dns/dyn/dyn.go:
```diff
@@ -0,0 +1,249 @@
+	Data json.RawMessage `json:"data"`
+
+	// The ID of the job that was created in response to a request.
+	JobId int `json:"job_id"`
```
This should be called `JobID`.

xenolf commented:
@aebruno the DNS tests seem to be a bit flaky on travis for some reason. I restarted the build and it went through OK.

I just added a few minor nitpicks from running `golint` on your code, other then that this is good to merge :)

aebruno commented:
@xenolf good catch. I made the dynRequest private to the package and cleaned up the JobID name. Thanks! Anything else just let me know.
